solutions/platinum/bts-hotcold.mdx

```diff
@@ -0,0 +1,185 @@
+
+In this case, we just update the path from $a$ to $p$ and
+$b$ to $p$. Note that if $p$ is equal to either $a$ or $b$,
+then this case must be handled slightly differently.
+
+**Case 2: $p$ is equal to either $a$ or $b$.**
+
+Let the first ancestor of $t$ that is on the path from $a$ to $b$
+be $l$. WLOG, let $b$ be equal to $p$. Then, we update the path from
+$a$ to $l$ and the path from $l$ to $b$.
+
+**Case 3: The LCA of $t$ with $a$ and $b$ is the same as $p$.**
+
+Handle this case similarly to case 1.
+
+**Case 4: Either the LCA of $t$ and $a$ or the LCA of $t$ and $b$ lie on the path from $a$ to $b$.**
+
+In this case, split the path updates into 3 segments. Let the lower of the two LCAs mentioned
+be $l$, and the node which is an ancestor of $l$ be $a$. Then, just update the path from
+$a$ to $l$, $l$ to $p$, and the path from $b$ to $p$.
+
+**Handling Path Updates**
+
+Note that based on how we did our casework on the paths, that all path updates are between a given
+node and its ancestor. Also, note that the path updates are just adding distances which either increase
+or decrease by a fixed amount every time. So, we can sort of do a difference array on a tree, where we DFS
+down the tree and apply the differences while walking back up the tree.
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+int main() {
+	int n, s;
+	cin >> n >> s;
+	// 1 indexing makes updates easier
+	vector<vector<int>> adj(n + 1);
+	for (int i = 1; i < n; i++) {
+		int u, v;
+		cin >> u >> v;
+		adj[u].push_back(v);
+		adj[v].push_back(u);
+	}
+
+	// perform Euler tour and binary lifts
+	const int LOG = 1 + (int)log2(n);
+	vector<vector<int>> lift(LOG, vector<int>(n + 1));
+	vector<int> tin(n + 1);
+	vector<int> tout(n + 1);
+	vector<int> depth(n + 1);
+	int timer = 0;
+	function<void(int, int)> tour = [&](int u, int p) {
+		tin[u] = timer++;
+		lift[0][u] = p;
+		depth[u] = depth[p] + 1;
+		for (int i = 1; i < LOG; i++) {
+			lift[i][u] = lift[i - 1][lift[i - 1][u]];
+		}
+		for (int v : adj[u]) {
+			if (v != p) { tour(v, u); }
+		}
+		tout[u] = timer - 1;
+	};
+	tour(1, 0);
+	tin[0] = -1, tout[0] = n + 1;  // so LCA works
+
+	// just some helper methods to simplify the code
+	auto isAnc = [&](int u, int v) -> bool {
+		return tin[u] <= tin[v] && tout[v] <= tout[u];
+	};
+	auto lca = [&](int u, int v) -> int {
+		if (isAnc(u, v)) return u;
+		if (isAnc(v, u)) return v;
+		for (int i = LOG - 1; i >= 0; i--) {
+			if (!isAnc(lift[i][u], v)) { u = lift[i][u]; }
+		}
+		return lift[0][u];
+	};
+	auto dist = [&](int a, int b, int anc = -1) -> int {
+		if (anc == -1) { anc = lca(a, b); }
+		return depth[a] + depth[b] - 2 * depth[anc];
+	};
+
+	// calculating difference array updates
+	vector<array<ll, 2>> diff(n + 1, {0, 0});
+	auto upd = [&](int u, int v, int initial, int change) -> void {
```

Comment:
at this point i'd rather you put things in a class to make things more encapsulated

Reply:
would this be like a Tree class where you put all the graph stuff in there?

Reply:
yeah, see my current pr for how it might go

Reply:
done